Lambda

aritorto opened the pull request.

goncalvesmachadoc committed:
addTestCaseSalt

akva2 committed:
changed: put ParallelOverlappingILU0 in separate compile unit

i chose to split in a separate _impl file because this code is so
generic that there may be downstream users who want to use on other
matrix types than what we use in opm-simulators.

goncalvesmachadoc committed:
deleta .data

GitPaean committed:
ignoring analytical aquifers without valid connections

aritorto committed:
Code modernization

Change from Functor class to functions.

blattms committed:
Merge pull request #4019 from akva2/paroverlap_separate_unit

put ParallelOverlappingILU0 in separate compile unit

akva2 committed:
changed: reduce some code duplication

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -25,6 +25,31 @@
+};
+
+//std::function<double(const double&)>
+static std::function<double(const double&)> auxiliaryFunctors::IdentityFunctor(const double& t){return t;};
```
You are returning a double and not a function. Maybe return the appropriate lambda expresssion?

aritorto replied:
I think it's fixed in the new iteration. (It's still running, no errors so far :))

aritorto replied:
Too optimistic! :/

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -25,6 +25,31 @@
+//std::function<double(const double&)>
+static std::function<double(const double&)> auxiliaryFunctors::SignFunctor(const double& t){if (t< 0){return -1;} else{return 1;}};
+//std::function<double(const double&)> 
+static std::function<double(const double&)> auxiliaryFunctors::IsPositiveFunctor(const double& t){if (t<0){return 0;} else{return 1;}};
```
return the lambda.

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -25,6 +25,31 @@
+//std::function<double(const double&)> 
+static std::function<double(const double&)> auxiliaryFunctors::OneFunctor(const double&){return 1.0;};
+//std::function<double(const double&)>
+static std::function<double(const double&)> auxiliaryFunctors::SignFunctor(const double& t){if (t< 0){return -1;} else{return 1;}};
```
return the lambda

blattms reviewed (commented):
I have not compiled your code and just pointed out some issues. Basically, you are saying that you return a function, but you are not.

Still it was very inspiring. Did not actually think of the possibility of a function returning a lambda, but that's :sunglasses:  :muscle: 

Actually now that I saw your code, we could actually make it even better by considering that a function is a std::function, too. Hence instead of the auxiliaryFunctors and lambdas, we could
also define free functions in the detail namespace and used them via std::function (but that is maybe for the next iteration):
```
namespace Opm::detail{
  double identityFunction(const double& t) {return t;}
}

void runFunction(std::function<double (const double&)> func, double value){
  auto val = func(value);
}
int main(){
  runFunc(identityFunction, 3);
}
```

aritorto commented:
It ran without errors now!
Let's me git add, commit, push again :)

I'll check your comment and reply in a sec!

aritorto commented:
I was a bit too enthusiastic.
I'll check again. There are still errors. I'll take into account your feedback!

blattms commented:
Maybe add/commit/push your changes and post the error messages. Without them it is rather hard to help you.

aritorto commented:
Yes, I'll do that

aritorto commented:
One error, for example, is:

```
/home/aritorto/Opm/opm-simulators/opm/simulators/linalg/MILU.cpp:260:14: error: ‘auxFunctors::AbsFunctor’ is not a type
  260 |     INSTANCE(auxFunctors::AbsFunctor,auxFunctors::SignFunctor,__VA_ARGS__) \
```
The part of the hpp file where I declared the functors and the part of the cpp file where they are called seem not to generate errors. But now, the INSTANCE part is not running. 
I added auxFunctors:: because of previous compiler-errors. I'll try one more time removing those extra auxFunctors::.

aritorto commented:
Now the definitions are in the cpp file!
Forgot to move them

blattms commented:
In the MILU.cpp file you need to change the defintion of mil0_decomposition to match the declaration (skip F1 and F2 and use std::function instead like in the header file).

Concerning the macros, it should probably read:
```
#define INSTANCE(...) \
    template void milu0_decomposition<__VA_ARGS__> \
       (__VA_ARGS__&, std::function<double(const double&)>, std::function<double(const double&)>, \
        std::vector<typename __VA_ARGS__::block_type>*);

#define INSTANCE_ILUN(...) \
    template void milun_decomposition(const __VA_ARGS__&, int, MILU_VARIANT, \
                                      __VA_ARGS__&,Reorderer&,Reorderer&);

#define INSTANCE_FULL(...) \
    INSTANCE(__VA_ARGS__) \
    INSTANCE_ILUN(__VA_ARGS__)

#define INSTANCE_BLOCK(Dim) \
    INSTANCE_FULL(Dune::BCRSMatrix<MatrixBlock<double,Dim,Dim>>)

#define INSTANCE_FM(Dim) \
    INSTANCE_FULL(Dune::BCRSMatrix<Dune::FieldMatrix<double,Dim,Dim>>)

INSTANCE_FM(1)
INSTANCE_FM(2)
INSTANCE_FM(3)
INSTANCE_FM(4)

INSTANCE_BLOCK(1)
INSTANCE_BLOCK(2)
INSTANCE_BLOCK(3)
INSTANCE_BLOCK(4)
INSTANCE_BLOCK(5)
INSTANCE_BLOCK(6)
```

aritorto commented:
Make sense!!!!!
Why would it be different from the declaration in the header file? :/

Thanks! Also the modification in the INSTANCE part makes sense...

aritorto commented:
I tried several things... But didn't manage to call correctly the functors when we use them as defualt arguments. 
Also, I wrote now functions (without the std::function) because I thought that was the new goal. 
But maybe I got it wrong. 
From what you wrote in my notebook, it should be:
```
template <typename T>
T oneFunctor(const T&);
```
inside the header...

Oh wait! I just realized I'm calling the functors in the cpp file as they were static and they are not static anymore. Should use the type class... Sorry. New commit soon!

I keep trying.

aritorto commented:
I'll add and commit the new version. Still with the same primary expression error.

From your notes in my notebook, it should be:
```
template <typename T>
T double someFunctor(const T&); // in the header, inside Opm, inside detail
```
```

template <typename T>
T Opm::detail::someFunctor(const T&){ return ...;}; // in the source "outside everything"
```

When I use them to define default arguments I tried many options:
1. keeping `std::function<double(const double&)> defaultFunctor1 = ...` or` double defualtFunctor1(const double&) = ...`
2.  I tried `Opm::detail<double> someFunctor() ` or `Opm::detail<double> someFunctor()(const double&) `as default argument.
Which might not work since we are inside Opm::detail so not sure we need those Opm::detail in the front. 
3. Also tried `someFunctor<Opm::detail><double>(const double&) `or `someFunctor<Opm::detail><double>()`
4. or `someFunctor<double>() `or `someFunctor<double>(const double&)`

Always getting thise type of error:
```
/home/aritorto/Opm/opm-simulators/opm/simulators/linalg/MILU.hpp: At global scope:
/home/aritorto/Opm/opm-simulators/opm/simulators/linalg/MILU.hpp:109:90: error: expected primary-expression before ‘<’ token
  109 | void milu0_decomposition(M& A, double absFunctor(const double&) = signFunctor<Opm::detail<double>>(const double&),
```

I won't give up! Just to keep you updated!

blattms commented on opm/simulators/linalg/MILU.cpp:
```diff
@@ -223,17 +232,17 @@ void milun_decomposition(const M& A, int n, MILU_VARIANT milu, M& ILU,
     case MILU_VARIANT::MILU_2:
-      detail::milu0_decomposition ( ILU, AuxFunctors::identityFunctor(),
-				    AuxFunctors::signFunctor()); 
+      detail::milu0_decomposition ( ILU, identityFunctor<double>(),
```
This works as it is only for double for now. But the actual type needs to be extracted from the Matrix class, please put
```
using field_type = typename M::field_type;
```
above the switch clause and then use e.g. `identityFunctor<field_type>` to refer to the function. Note: without bracketsas we are not calling the function but want to pass it. Actually you may use the complete signature (with function parameters), but that would be `identityFunctor<double>(const double&)`, wouldn't it? There is no `identityFunctor` without arguments.

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -86,14 +106,14 @@ struct RealReorderer : public Reorderer
 template<class M>
-void milu0_decomposition(M& A, std::function<double(const double&)> absFunctor = AuxFunctors::signFunctor(),
-			  std::function<double(const double&)> signFunctor = AuxFunctors::oneFunctor(),
+void milu0_decomposition(M& A, double absFunctor(const double&) = signFunctor<Opm::detail<double>>(const double&),
+			 double signFunctor(const double&) = oneFunctor<Opm::detail<double>>(const double&),
```
This seems a bit messed up. You want to pass it the signFunctor function in the Opm::detail namespace as default arguments.  As you are in ` Opm::detail` the namespace can be omitted here. Please be aware that `Opm::detail<double>` is wrong. It may be double, but more correct would be to use `typename M::field_type` instead. Also to refer to a function you probably may omit the brackets with function parameters but they should not hurt either

But the type of the function parameter does not change it is still a function that gets a double and returns a double like before, hence `std::function<double(const double&)> `. Hence no changes to the old version before the `=` signs, just after them.

aritorto replied:
New version std::function are back! So far no errors but compilation hasn't finished yet
Thakns!
sorry for the many confusing pieces of code :

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -86,14 +106,14 @@ struct RealReorderer : public Reorderer
 void  milu0_decomposition(M& A, std::vector<typename M::block_type>* diagonal)    
  {
-   milu0_decomposition(A, AuxFunctors::identityFunctor(), AuxFunctors::oneFunctor(), diagonal);
+   milu0_decomposition(A, identityFunctor<Opm::detail<double>>(const double&), oneFunctor<Opm::detail<double>>(const double&), diagonal);
```
dito

aritorto replied:
new version with std::functions is compiling, so far with no errors.... let's see 
PS: I also just leant what dito is :)

blattms replied:
that is of course elementary/vital knowledge, How you survived without it before is beyond me. :laughing: :rofl:

blattms reviewed (commented):
Just commented but did not try to compile, yet.

blattms commented:
Well, I cannot refer to function with an argument list. Only something like this works (not sure why):
```
std::function<double (const double&)> = signFunctor<double>;
```

Maybe for now just make it compile for double (skipping `typename M::field_type` as this would also need to be used for std::function):
```
std::function<typename M::field_type (const typename M::field_type&)> = signFunctor<double>;
```
You would also need to use it for the instance thingies.

aritorto commented:
There are erros coming from the other file Parallel[...] and many error of basically these two types:

```
/home/aritorto/Opm/opm-simulators/opm/simulators/linalg/ParallelOverlappingILU0.hpp:790:49: error: cannot resolve overloaded function ‘signFunctor’ based on conversion to type ‘std::function<double(const double&)>’
  790 |                     detail::milu0_decomposition ( *ILU);
```
```

/home/aritorto/Opm/opm-simulators/opm/simulators/linalg/MILU.cpp:240:35: error: no matching function for call to ‘milu0_decomposition(Dune::BCRSMatrix<Dune::FieldMatrix<double, 1, 1> >&, <unresolved overloaded function type>, <unresolved overloaded function type>)’
  240 |       detail::milu0_decomposition ( ILU, absFunctor<double>,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
  241 |         signFunctor<double>);
```

I'll continue tomorrow.

blattms commented on opm/simulators/linalg/MILU.cpp:
```diff
@@ -33,19 +33,19 @@
 
 template <typename T>
-T Opm::detail::absFunctor(const T& t){return std::abs(t);};
+std::function<T(const T&)>  Opm::detail::absFunctor(const T& t){return std::abs(t);};
```
This was correct before. The return type has to match what you return. Please check what you return (in the new version you are saying a `std::function<T(const T&)> but the function body says otherwise.

aritorto replied:
Checked! 
Now templates are a bit mixed-type, in the sense that `oneFunctor`, `signFunctor`, and `isPositiveFunctor` return double and take `const T&`

I'm compiling a second time because there was an ambiguous thing between my template declarations in the header file and template definitions in the source file.

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -59,24 +59,24 @@ MILU_VARIANT convertString2Milu(const std::string& milu);
-  T absFunctor(const T&);
-  // std::function<T(const T&)> absFunctor();
+  // T absFunctor(const T&);
+  std::function<T(const T&)> absFunctor();
```
IMHO the original version was correct. Those functions take a `const double&` as an argument and return a double, right.

aritorto replied:
Yes, got it now... see one of my comments below where I explained my confusion about `std::function` in the front of my declaration :/
Thanks!

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -106,14 +106,14 @@ struct RealReorderer : public Reorderer
 void  milu0_decomposition(M& A, std::vector<typename M::block_type>* diagonal)    
  {
-   milu0_decomposition(A, identityFunctor<Opm::detail<double>>(const double&), oneFunctor<Opm::detail<double>>(const double&), diagonal);
+   milu0_decomposition(A, identityFunctor<double>, oneFunctor<double>, diagonal);
```
:+1:  This is what I meant.

blattms commented on opm/simulators/linalg/MILU.cpp:
```diff
@@ -255,11 +255,11 @@ void milun_decomposition(const M& A, int n, MILU_VARIANT milu, M& ILU,
+  template std::function<double(const double&)> oneFunctor;
+  template std::function<double(const double&)> signFunctor;
+  template std::function<double(const double&)> isPositiveFunctor;
+  template std::function<double(const double&)> absFunctor;
```
Seemed correct before

aritorto replied:
So even if the functions are declared in `Opm::detail `as **non-static** functions, they can/must be called with the `Opm::detail:: `in front. 
I thought that was only for **static member functions** and "the **non-static**" ones should be called like `myFuction<myClass::mySubClass><type_if_it_comes_from_a_template>`.

I was wrong :/

blattms commented on opm/simulators/linalg/MILU.cpp:
```diff
@@ -232,17 +232,17 @@ void milun_decomposition(const M& A, int n, MILU_VARIANT milu, M& ILU,
-      detail::milu0_decomposition ( ILU, identityFunctor<double>() ,
-				    isPositiveFunctor<double>());
+      detail::milu0_decomposition ( ILU, identityFunctor<double> ,
+				    isPositiveFunctor<double>);
```
Perfect. :+1:

aritorto replied:
So here, as you mentioned yesterday, the last two arguments of `milu0_decomposition` are functions that receive double and return `const double& `type objects. So the (only) the name of those functions(with the type double since they were defined in a `template <typename T>`) should be the argument, **without the extra ()**.
Thanks again!

blattms commented on opm/simulators/linalg/MILU.hpp:
```diff
@@ -43,8 +45,38 @@ enum class MILU_VARIANT{
+  static std::function<double(const double&)> absFunctor(); 
+  };*/
+  
+  template <typename T>
+  //T identityFunctor(const T&);
+  std::function<T(const T&)> identityFunctor();
+
+  template <typename T>
+  //T oneFunctor(const T&);
+  std::function<T(const T&)> oneFunctor();
+
+  template <typename T>
+  //T signFunctor(const T&);
+  std::function<T(const T&)> signFunctor();
+
+  template <typename T>
+  //T isPositiveFunctor(const T&);
+  std::function<T(const T&)> isPositiveFunctor();
+
+  template <typename T>
+  // T absFunctor(const T&);
+  std::function<T(const T&)> absFunctor();
```
Not sure why you changed that. Seems like the previous version was perfectly ok.
Those functions take a double parameter and return a double. With this change you are now saying that they take no parameter and return a std::function.

aritorto replied:
Somehow I got confused about the` std::function<type1(type2)>` in the front. Thinking that I'm defining/declaring a function that takes type2 objects and returns type1 objects. But there, I'm saying that `absFunctor()` returns a `std::function<type1(type2)>` object. Now, I see the difference/error. Thanks!

blattms reviewed (commented):
Seems like I managed to confuse you big time again.

I meant to use `std::function<double(const doubele&)` as the type for only the parameters of `milu0_decomposition` and not anywhere else. The rest of the changes confused the compiler because it searched for functions that take a `const double&` and return a `double` with the names `signFunctor`, etc. and could not find such overloads. That caused these errors.

I will post a PR to your branch to clarify things and we should discuss tomorrow morning.

bska committed:
Merge pull request #4002 from goncalvesmachadoc/testSalt

add Test Case for Salt Precipitation

aritorto commented:
Thanks a lot for your corrections!
Now it's compiling, finger crossed...

aritorto commented:
Compilation error 0!!!!!!!!!!!!!!!!!!!!!
What a great beginning of the day! I'll git add, commit, push...

akva2 committed:
changed: get rid of duplicate MatrixBlock headers/classes

this has already led to some confusion. move some of the code
upstream to opm-models and remove the rest of the duplicated code.

the remainder of MatrixBlock.hpp is renamed to SmallDenseMatrixUtils.hpp

bska committed:
Merge pull request #4015 from GitPaean/handling_numerical_aquifer_no_connection

ignoring analytical aquifers without valid connections

atgeirr committed:
Add packing support for classes needing it.

Since Rates is not longer a POD, it needs explicit
packSize()/pack()/unpack() support. This also applies
to classes containing Rates, namely Connection and Segment.
Similarly, SegmentPressures is not a POD.

atgeirr committed:
Add override qualifier to silence warning.

blattms commented on opm/simulators/linalg/ParallelOverlappingILU0.hpp:
```diff
@@ -37,6 +37,21 @@
 #include <cstddef>
 #include <string>
 
+////////////////////////////////////////////////////
+
+
+  
+
+
+
+
+
+
+
+
+
+///////////////////////////////////////////////////
+
```
This should be removed, too.

blattms reviewed (commented):
Wow,this looks really nice now. :muscle: :sunglasses: Great work. We should tap ourselves on the shoulder for it.

There is still one comment to remove. Please remove that now.

After that we should probably make this one commit and fix indentation (probably in a new branch). I will try to give you details on this via email.

atgeirr committed:
Merge pull request #4024 from atgeirr/silence-override-warning

Add override qualifier to silence warning.

aritorto commented:
Done!
Unfortunately in 2 commits. Forgot one comment in the first commit.

bska committed:
Merge pull request #4023 from atgeirr/fix-packing

Add packing support for classes needing it.

bska committed:
Merge pull request #4022 from akva2/reduce_dup

changed: reduce some code duplication

bska committed:
Prefer Attributes to Macros

bska committed:
Revert "changed: reduce some code duplication"

bska committed:
Merge pull request #4026 from OPM/revert-4022-reduce_dup

Revert "changed: reduce some code duplication"

atgeirr committed:
Merge pull request #4025 from bska/remove-dune-unused

Prefer Attributes to Macros

blattms commented:
Closing this as there is an official one now.
